specs/fizzbee/LeaderFollower/VeniceLeaderFollower.fizz

```diff
@@ -0,0 +1,136 @@
+    NODE_IDS = range(N_NODES)
+
+    # Initialize the state of each node
+    real_time_topic = []
+    version_topic = []
+    nodes = [
+        {
+            'state': 'FOLLOWER',
+            'rtOffset': 0,
+            'vtOffset': 0,
+            'persistedRecords': {}
+        }
+        for i in NODE_IDS
+    ]
+
+
+# The action keyword is used to define an action. The atomic keyword implies
+# there are no yield points within the statements.
+atomic action PromoteLeader:
+    # A FOLLOWER ca be promoted to be a LEADER if and only if there is no leader in the system.
+
+    if [node for node in nodes if node['state'] == 'LEADER']:
+        return
+    # `any` is a special keyword that creates non-determinism or alternate timelines.
+    # This is similar in structure to `for`, but in each timeline, only one of the
+    # non-deterministic choices is executed.
+    any id in NODE_IDS:
+        nodes[id]['state'] = 'LEADER'
+
+
+atomic action DemoteLeader:
+    any id in NODE_IDS:
+        if nodes[id]['state'] == 'LEADER':
+            nodes[id]['state'] = 'FOLLOWER'
+
+
+atomic action ClientProducesToVenice:
+    any kv in [(k, v) for k in KEYS for v in VALUES]:
+        real_time_topic.append(kv)
+
+
+# The fair keyword implies this action is weakly fair. That is, the action is always enabled,
+# it will eventually happen.
+atomic fair action LeaderConsume:
+  any id in NODE_IDS:
+    if nodes[id]['state'] == 'LEADER':
+        if nodes[id]['vtOffset'] >= len(version_topic):
+            real_time_consume(id)
+        else:
+            version_topic_consume(id)
+
+
+# The strong modifier to the fair keyword implies this action is strongly fair. That is, the action
+# gets enabled repeatedly (that is it may get disabled in between), it will eventually happen.
+atomic fair<strong> action FollowerConsume:
+    any id in NODE_IDS:
+        if nodes[id]['state'] == 'FOLLOWER' and nodes[id]['vtOffset'] < len(version_topic):
+            version_topic_consume(id)
+
+
+atomic func real_time_consume(nodeId):
+    # A leader consuming unprocessed msgs from the real-time topic.
+
+    # Check if the node's offset is within the bounds of the real-time topic
+    if nodes[nodeId]['rtOffset'] >= len(real_time_topic):
+        return
+
+    # Update the node's state and persisted records
+    offset = nodes[nodeId]['rtOffset']
+    key, value = real_time_topic[offset][0], real_time_topic[offset][1]
+
+    nodes[nodeId]['rtOffset'] += 1
+    nodes[nodeId]['persistedRecords'][key] = value
+
+    # Append the consumed event to the version topic
+    version_topic.append((key, value, offset))
+
+
+atomic func version_topic_consume(nodeId):
+    # A node consuming events from the version topic.
+
+    # Update the node's state and persisted records
+    offset = nodes[nodeId]['vtOffset']
+    key, value = version_topic[offset][0], version_topic[offset][1]
+    nodes[nodeId]['vtOffset'] += 1
+    nodes[nodeId]['persistedRecords'][key] = value
```

Comment:
In a similar vein as above, we do the persistence first, and the VT offset increment (i.e. checkpointing) after, so that we tolerate crashing in-between the two operations (and we may want to remove `atomic`...).

Reply:
Same as above response, because the model here is very simplistic (high level?) the line ordering here doesn't matter too much.  It'd be better to split each of these into their own states.